README.md
# Vase // Elasticsearch

[![Coverage Status](https://coveralls.io/repos/github/nytimes/vase.elasticsearch/badge.svg?branch=master&t=aFiCSQ)](https://coveralls.io/github/nytimes/vase.elasticsearch?branch=master)

Elasticsearch extensions for [Vase][vase] and [Fern][fern].

This library adds a few literals which make it possible to quickly
create APIs using Elasticsearch. This uses the Elasticsearch library
[Spandex][spandex] for the underlying Elasticsearch client.

## Usage
### vase.elasticsearch/connect

This literal will connect to an Elasticsearch and return an
interceptor which injects the connection in the context on every
request.

The only required argument is `:hosts` which accepts a collection of
URIs of nodes. Additional allowed arguments specified [here][spandex-client]

#### Example

```clojure
(fern/lit vase.elasticsearch/connect
          {:hosts ["http://localhost:9200"]})
```

### vase.elasticsearch/create

Create an Elasticsearch index using the provided name and settings. If
an index with the same name exists, this is a no-op.

| Param       | Meaning                                                            |
|:------------|:-------------------------------------------------------------------|
| `:name`     | Optional. The name of the interceptor.                             |
| `:index`    | The name of the index.                                             |
| `:settings` | The index settings specified inline, in a JSON, or in a YAML file. |

#### Example

This creates an index named `tmdb` using the Elasticsearch
configuration specified in the file `settings.json`.

```clojure
(fern/lit vase.elasticsearch/create
          {:name     :tmdb/create
           :index    "tmdb"
           :settings "settings.json"})
```

### vase.elasticsearch/get-document

Get a document from the index by its id.

| Param      | Meaning                                                            |
|:-----------|:-------------------------------------------------------------------|
| `:name`    | Optional. The name of the interceptor.                             |
| `:index`   | The name of the index.                                             |
| `:type`    | Optional. The document type. Defaults to `:_doc`                   |
| `:id-path` | The path within the [context map][context-map] to the document id. |

#### Example

This builds an interceptor which will lool up a document of type
`:_doc` in the index `my-index`. The id is sourced from the path
paramter `:id`.

``` clojure
(fern/lit vase.elasticsearch/get-document
                           {:name    :tmdb/get-document
                            :index   "my-index"
                            :type    :_doc
                            :id-path [:request :path-params :id]})
```

### vase.elasticsearch/put-document

Put a document into the index.

| Param        | Meaning                                                                                                                |
|:-------------|:-----------------------------------------------------------------------------------------------------------------------|
| `:name`      | Optional. The name of the interceptor.                                                                                 |
| `:index`     | The name of the index.                                                                                                 |
| `:type`      | Optional. The document type. Defaults to `:_doc`.                                                                                          |
| `:id-path`   | The path within the [context map][context-map] to the document id.                                                     |
| `:body-path` | The path within the [context map][context-map] to the document body. This is what will get indexed into Elasticsearch. |

#### Example

``` clojure
(fern/lit vase.elasticsearch/put-document
                           {:name      :tmdb/put-document
                            :index     "my-index"
                            :type      :_doc
                            :id-path   [:request :path-params :id]
                            :body-path [:request :body]})
```

### vase.elasticsearch/delete-document

Delete a document from the index by its id.

| Param      | Meaning                                                            |
|:-----------|:-------------------------------------------------------------------|
| `:name`    | Optional. The name of the interceptor.                             |
| `:index`   | The name of the index.                                             |
| `:type`    | Optional. The document type.                                       |
| `:id-path` | The path within the [context map][context-map] to the document id. |

#### Example

``` clojure
(fern/lit vase.elasticsearch/delete-document
                           {:name    :tmdb/delete-document
                            :index   "my-index"
                            :type    :_doc
                            :id-path [:request :path-params :id]})
```

### vase.elasticsearch/search

Perform a search against an Elasticsearch index (or indices if a
vector of index names). See the [Elasticsearch docs][search-docs] for
more details.

| Param     | Meaning                                                                                                                                             |
|:----------|:----------------------------------------------------------------------------------------------------------------------------------------------------|
| `:name`   | Optional. The name of the interceptor.                                                                                                              |
| `:index`  | The name of the index.                                                                                                                              |
| `:method` | Optional. The HTTP method to use when querying Elasticsearch. Defaults to `:get`.                                                                   |
| `:params` | Variables to bind from the request map. These values come from query and path parameters.                                                           |
| `:body`   | The query to send to Elasticsearch. You may use programming constructs within the body. See the [Elasticsearch docs][search-docs] for more details. |

#### Example

```clojure
(fern/lit vase.elasticsearch/search
          {:name   :tmdb/count
           :index  "my-index"
           ;; The variable `q` comes a query paramter. So, e.g., if we
           ;; have some endpoint that's queryied like so
           ;; `/search?q=nytimes`, `q` would be bound to `nytimes`.
           :params [q]
           ;; Notice that we are using a programming construct within
           ;; an otherwise declarative query. This opens up a wide
           ;; possibility of behavior within a compact form.
           :body   {:query
                    (if q
                      {:multi_match {:query  q
                                     :fields ["overview"
                                              "title"
                                              "directors.name"
                                              "cast.name"]}}
                      {:match_all {}})}})
```


### vase.elasticsearch/count

Return a count for the number of matches for a given query. See the
[Elasticsearch docs][count-docs] for more details.

| Param     | Meaning                                                                                                                                                  |
|:----------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|
| `:name`   | Optional. The name of the interceptor.                                                                                                                   |
| `:index`  | The name of the index.                                                                                                                                   |
| `:method` | Optional. The HTTP method to use when querying Elasticsearch. Defaults to `:get`.                                                                        |
| `:params` | Variables to bind from the request map. These values come from query and path parameters.                                                                |
| `:body`   | The count query to send to Elasticsearch. You may use programming constructs within the body. See the [Elasticsearch docs][count-docs] for more details. |

### vase.elasticsearch/request

Perform an abritrary HTTP request against Elasticsearch. You can use
this to peform any request not covered by one of the named actions above.

| Param     | Meaning                                                                                   |
|:----------|:------------------------------------------------------------------------------------------|
| `:name`   | Optional. The name of the interceptor.                                                    |
| `:index`  | The name of the index.                                                                    |
| `:method` | Optional. The HTTP method to use when querying Elasticsearch. Defaults to `:get`.         |
| `:params` | Variables to bind from the request map. These values come from query and path parameters. |
| `:body`   | The query to send to Elasticsearch. You may use programming constructs within the body.   |

#### Constructing a query

```clojure
(fern/lit vase.elasticsearch/request
          {:params [q dbg]
           :url    [:tmdb :_search]
           :method :get
           :body   {:_source [:title :release_year]
                    :explain (or dbg false)
                    :query   (let [q (or q "")]
                               {:multi_match
                                {:query  q
                                 :fields ["title^0.1", "overview"]}})}})

```

#### Indexing content

```clojure
(fern/lit vase.elasticsearch/request
          {:params [id]
           :url    [:tmdb :_doc id]
           :method :put})

```

## Running the Example

Start an Elasticsearch cluster with Docker compose:

```sh
docker-compose up
```

In a new terminal, start the example API:

```sh
clj -A:test:service
```

In a new terminal, index content:

```sh
clj -A:test:index
```

You can now start issuing queries against the API:

```sh
curl -s 'localhost:8080/search?q=basketball+with+cartoon+aliens' | jq .hits.hits[]._source.title
"Space Jam"
"Just Wright"
"Who Wants To Kill Jessie?"
"He Got Game"
"Grown Ups"
"Air Bud"
"Glory Road"
"Semi-Pro"
"Coach Carter"
"Speed Racer"
```

The entire API is define using the configuration file in
`dev/tmdb.fern`. You can easily iterate on this query to make a search
that works for you.

Relasing
--------

Example:

``` sh
$ make release/patch clean jar deploy
```

[vase]: https://github.com/cognitect-labs/vase
[fern]: https://github.com/cognitect-labs/fern
[spandex]: https://github.com/mpenet/spandex
[spandex-client]: https://mpenet.github.io/spandex/qbits.spandex.html#var-client
[context-map]: http://pedestal.io/reference/context-map
[search-docs]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-search.html
[count-docs]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-count.html


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - TBD

Initial release.

[Unreleased]: https://github.com/olivierlacan/keep-a-changelog/compare/1.0.0...HEAD
[1.0.0]: https://github.com/olivierlacan/keep-a-changelog/releases/tag/1.0.0


test/com/nytimes/vase/fixture.clj
(ns com.nytimes.vase.fixture
  "Elasticsearch test fixture."
  (:require [clj-test-containers.core :as test-container]))

(defonce ^{:doc "Reference to the Elasticsearch cluster."}
  container nil)

(defonce ^{:doc "Reference to the Elasticsearch client."}
  client nil)

(def container-definition
  {:image-name    "docker.elastic.co/elasticsearch/elasticsearch:7.5.2"
   :exposed-ports [9200]
   :env-vars      {"cluster.name"          "test-cluster"
                   "bootstrap.memory_lock" "true"
                   "discovery.type"        "single-node"
                   "ES_JAVA_OPTS"          "-Xms512m -Xmx512m"}})

(defn start-elasticsearch []
  (alter-var-root #'container
                  (constantly
                   (test-container/start!
                    (test-container/create container-definition)))))

(defn stop-elasticsearch []
  (alter-var-root #'container test-container/stop!))

(defn start-client []
  (alter-var-root #'client
                  (constantly
                   {:hosts [(format "http://%s:%d"
                                    (get-in container [:host])
                                    (get-in container [:mapped-ports 9200]))]})))

(defn elasticsearch-fixture [test-fn]
  (start-elasticsearch)
  (start-client)
  (test-fn)
  (stop-elasticsearch))




test/com/nytimes/vase/elasticsearch_test.clj
(ns com.nytimes.vase.elasticsearch-test
  (:require [clojure.java.io :as io]
            [clojure.test :as test :refer [deftest is testing]]
            [com.nytimes.vase.elasticsearch :as vase.es]
            [fern :as f]
            [io.pedestal.interceptor :as i]
            [io.pedestal.interceptor.chain :as chain]
            [com.nytimes.vase.fixture :refer [elasticsearch-fixture client]]
            [qbits.spandex :as es]))

(def index
  (str (gensym "test-index-")))

(defn context-execute
  "Execute the interceptor chain with the ES client already inject into the
  context."
  ([interceptors]
   (context-execute {} interceptors))
  ([extra interceptors]
   (chain/execute (merge {::vase.es/client (es/client client)} extra)
                  (map i/interceptor interceptors))))

(defn absolute-path [path]
  (.getAbsolutePath (io/file path)))

(defn ok? [context]
  (let [status (-> context :response :status)]
    (<= 200 status 299)))

(defn response
  "Extract the response from the interceptor context."
  [context]
  (-> context :response (dissoc :hosts)))

;; ----------------------------------------------------------------------
;; Tests

(test/use-fixtures :once elasticsearch-fixture)

(deftest connect
  (is (i/interceptor? (f/literal 'vase.elasticsearch/connect client))))

(deftest content-type
  (is (= "application/json" (vase.es/content-type "test.json")))
  (is (= "application/yaml" (vase.es/content-type "test.yaml")))
  (is (= "application/yaml" (vase.es/content-type "test.yml"))))

(deftest create
  (testing "json"
    (is (= {:acknowledged true, :shards_acknowledged true, :index "json-test"}
           (:response (context-execute
                       [(f/literal 'vase.elasticsearch/create
                                   {:index     "json-test"
                                    :settings (.getFile (io/resource "com/nytimes/vase/settings.json"))})])))))

  (testing "yaml"
    (is (= {:acknowledged true, :shards_acknowledged true, :index "yaml-test"}
           (:response (context-execute
                       [(f/literal 'vase.elasticsearch/create
                                   {:index     "yaml-test"
                                    :settings (.getFile (io/resource "com/nytimes/vase/settings.yaml"))})])))))

  (testing "direct"
    (is (= {:acknowledged true, :shards_acknowledged true, :index "direct-test"}
           (:response (context-execute
                       [(f/literal 'vase.elasticsearch/create
                                   {:index     "direct-test"
                                    :settings {:settings
                                               {:index
                                                {:number_of_shards   1
                                                 :number_of_replicas 0}}
                                               :mappings
                                               {:properties
                                                {:foo {:type :keyword}}}}})]))))))


(deftest CRUD
  (testing "create-index"
    (context-execute
     [(f/literal 'vase.elasticsearch/create
                 {:index    index
                  :settings {:settings
                             {:index
                              {:number_of_shards   1
                               :number_of_replicas 0}}
                             :mappings
                             {:properties
                              {:foo {:type :keyword}}}}})]))

  (testing "put-document"
    (let [documents [{:id 1, :body {:foo "bar"}},
                     {:id 2, :body {:foo "baz"}},
                     {:id 3, :body {:foo "quux"}}]]
      (doseq [document documents]
        (testing (str document)
          (is (ok? (context-execute
                    document
                    [(f/literal 'vase.elasticsearch/put-document
                                {:name      :crud/put-document
                                 :index     index
                                 :id-path   [:id]
                                 :body-path [:body]})]))))))
    ;; we do this to make sure that the documents we just indexed are
    ;; visible to the following operations.
    (Thread/sleep 5000))
  
  (testing "get-document"
    (is (= {:_id           "1"
            :_index        index
            :_primary_term 1
            :_seq_no       0
            :_source       {:foo "bar"}
            :_type         "_doc"
            :_version      1
            :found         true}
           (:body
            (response
             (context-execute
              {:request {:path-params {:id 1}}}
              [(f/literal 'vase.elasticsearch/get-document
                          {:name    :crud/get-document
                           :index   index
                           :id-path [:request :path-params :id]})]))))))

  (testing "update-document"
    (let [documents [{:id 1, :body {:foo "bar1"}},
                     {:id 2, :body {:foo "baz1"}},
                     {:id 3, :body {:foo "quux1"}}]]
      (doseq [document documents]
        (testing document
          (is (ok? (context-execute
                    document
                    [(f/literal 'vase.elasticsearch/put-document
                                {:name      :crud/update-document
                                 :index     index
                                 :id-path   [:id]
                                 :body-path [:body]})])))))))

  (testing "search"
    (is (= [{:_id "1", :_index index, :_score 1.0, :_source {:foo "bar"}, :_type "_doc"}
            {:_id "2", :_index index, :_score 1.0, :_source {:foo "baz"}, :_type "_doc"}
            {:_id "3", :_index index, :_score 1.0, :_source {:foo "quux"}, :_type "_doc"}]
           (get-in
            (response
             (context-execute
              [(f/literal 'vase.elasticsearch/search
                          {:name  :search/search
                           :index index
                           :body  {:query {:match_all {}}}})]))
            [:body :hits :hits]))))

  (testing "count"
    (is (= {:_shards {:failed 0, :skipped 0, :successful 1, :total 1}, :count 3}
           (:body
            (response
             (context-execute
              [(f/literal 'vase.elasticsearch/count
                          {:name  :search/count
                           :index index
                           :body  {:query {:match_all {}}}})]))))))

  (testing "bad query"
    (is (=  {:body {:error  {:col        19,
                             :line       1,
                             :reason     "[terms] query malformed, no start_object after query name",
                             :root_cause [{:col    19,
                                           :line   1,
                                           :reason "[terms] query malformed, no start_object after query name",
                                           :type   "parsing_exception"}],
                             :type       "parsing_exception"},
                    :status 400},
             :exception-type :clojure.lang.ExceptionInfo,
             :interceptor :search/count,
             :stage       :enter,
             :status      400,
             :type        :qbits.spandex/response-exception}
           (select-keys
            (response
             (context-execute
              [(f/literal 'vase.elasticsearch/count
                          {:name  :search/count
                           :index index
                           :body  {:query {:terms ["baz1" "quux1"]}}})]))
            [:body :exception-type :type :status :stage :interceptor]))))

  (testing "bad request"
    (is (= {:body           {:error  (str "Incorrect HTTP method for uri [/"
                                          index
                                          "/fake-type] and method [GET], allowed: [POST]")
                             :status 405}
            :exception-type :clojure.lang.ExceptionInfo
            :interceptor    :crud/flush
            :stage          :enter
            :status         405
            :type           :qbits.spandex/response-exception}
           (select-keys 
            (response
             (context-execute
              [(f/literal 'vase.elasticsearch/request
                          {:name   :crud/flush
                           :method :get
                           :url    [index :fake-type]})]))
            [:body :exception-type :type :status :stage :interceptor]))))

  (testing "delete-document"
    (let [documents [{:id 1}, {:id 2}, {:id 3}]]
      (doseq [document documents]
        (testing document
          (is (ok? (context-execute
                    document
                    [(f/literal 'vase.elasticsearch/delete-document
                                {:name    :crud/delete-document
                                 :index   index
                                 :id-path [:id]})]))))))))


src/com/nytimes/vase/elasticsearch.clj
(ns com.nytimes.vase.elasticsearch
  (:require [clojure.core.match :refer [match]]
            [clojure.java.io :as io]
            [clojure.spec.alpha :as s]
            [clojure.string :as str]
            [com.cognitect.vase.actions :as actions]
            [com.cognitect.vase.fern :as vf]
            [fern :as f]
            [io.pedestal.interceptor :as i]
            [io.pedestal.interceptor.helpers :as interceptor]
            [io.pedestal.log :as log]
            [qbits.spandex :as es])
  (:import clojure.lang.ExceptionInfo
           java.io.Writer
           [org.elasticsearch.client Node RestClient]))

(s/def :vase.elasticsearch.index/name     string?)
(s/def :vase.elasticsearch.index/settings (s/or :file-name string? :edn map?))

(defn hosts
  [^RestClient client]
  (into [] (for [^Node node (.getNodes client)]
             (str \" (.getHost node) \"))))

(defmethod print-method RestClient
  [^RestClient client ^Writer writer]
  (.write writer (format "org.elasticsearch.client.RestClient[%s]"
                         (str/join "," (hosts client)))))


(defn error-response
  "Extract the error response from an Elasticsearch exception."
  [^clojure.lang.ExceptionInfo ex]
  (-> ex ex-data))

(defn error-status
  [exception]
  (let [data (ex-data exception)]
    (get data :status 500)))

(defn error-body
  [exception]
  (let [data (ex-data exception)]
    (get-in data [:body :error] data)))

(def default-type
  "This is the default document type in Elasticsearch."
  :_doc)

;; ----------------------------------------------------------------------
;; Connect

(defn ready?
  [client]
  (log/info :msg "Attempting to connect to Elasticsearch." :client client)
  (try
    (es/request client
                {:method       :get
                 :url          "_cluster/health"
                 :query-string {"wait_for_status" "yellow"}})
    true
    (catch Throwable _ false)))

(defmethod f/literal 'vase.elasticsearch/connect
  [_ opts]
  (let [client (es/client opts)]
    (while (not (ready? client))
      (Thread/sleep 1000))
    (log/info :msg "Adding Elasticsearch client.")
    (interceptor/before
     ::connect
     (fn [context]
       (assoc context ::client client)))))

;; ----------------------------------------------------------------------
;; Generic Elasticsearch Request

(defn merged-parameters
  "Create a map of all possible parameters by extracting and then
  merging them from the request object."
  [{:keys [query-params path-params params json-params edn-params]}]
  {:post [(map? %)]}
  (merge (if (empty? path-params) {} (actions/decode-map path-params))
         params
         json-params
         edn-params
         query-params))

(defn request-expr
  "Build a function which queries Elasticsearch and returns the response."
  [{:keys [params url method body]}]
  (assert (some? url) "URL must be a string or a vector of URL components")
  (assert (some? method) "Method must be a valid HTTP method such as :get or :post")
  `(fn [{~'request :request :as ~'context}]
     (let [client#                (::client ~'context)
           ~(actions/bind params) (merged-parameters ~'request)
           request#               {:url    ~url
                                   :method ~method
                                   :body   (or ~body (:body ~'request))}]
       (assoc ~'context :response (es/request client# request#)))))

(defrecord ElasticsearchRequestAction [name params url method body]
  i/IntoInterceptor
  (-interceptor [_]
    (actions/dynamic-interceptor
     name
     {:enter
      (request-expr {:url    url
                     :method method
                     :params params
                     :body   body})

      :error
      (fn [context exception]
        (log/error :msg "Elasticsearch request failed"
                   :status (error-status exception)
                   :error (error-body exception))
        (assoc context :response (error-response exception)))})))

(defmethod f/literal 'vase.elasticsearch/request [_ d] (map->ElasticsearchRequestAction (vf/with-name d)))

;; ----------------------------------------------------------------------
;; Indices API

(defn index-exists?
  "Return true if name index exists."
  [client index]
  (= 200 (try
           (:status (es/request client {:url index :method :head}))
           (catch Exception e
             (:status (ex-data e))))))

(defn content-type
  "Return settings file content type."
  [file-name]
  {:pre [(some? file-name) (not (str/blank? file-name))]}
  (if (re-find #"\.ya?ml$" file-name)
    "application/yaml"
    "application/json"))

(s/def :vase.elasticsearch/create
  (s/keys :req-un [:vase.elasticsearch.index/name
                   :vase.elasticsearch.index/settings]))


;; Create:

(defrecord CreateIndexAction [name index settings]
  i/IntoInterceptor
  (-interceptor [_]
    (assert (and (string? index) (not (str/blank? index)))
            "Index must be a non-empty string")
    (assert (or (map? settings)
                (and (string? settings)
                     (or (str/ends-with? settings ".json")
                         (str/ends-with? settings ".yaml")
                         (str/ends-with? settings ".yml"))))
            "Settings must be a map or a file ending with .json, .yaml, or .yml")
    (actions/dynamic-interceptor
     name
     {:enter
      (fn [{:keys [::client] :as context}]
        (when (nil? client)
          (throw (ex-info (str "No Elasticsearch client in the context. "
                               "Is the vase.elasticsearch/connect interceptor in the interceptor chain?")
                          {})))

        (if (index-exists? client index)
          (do (log/info :msg "Index already exists, doing nothing") context)
          (let [request  {:url     index
                          :method  :put
                          :headers {"content-type" (if (string? settings)
                                                     (content-type settings) "application/json")
                                    "accept"       "application/json"}
                          :body    (if (string? settings)
                                     (es/raw (slurp (io/file settings)))
                                     settings)}
                response (es/request client request)
                status   (:status response)
                body     (:body response)]
            (log/info :msg "Index created" :status status :response body)
            (assoc context :status status :response body))))

      :error
      (fn [context exception]
        (log/error :msg "Failed to create index. See stracktrace for Elasticsearch error."
                   :status (error-status exception)
                   :error (error-body exception))
        (throw exception))})))

;; Literals:

(defmethod f/literal 'vase.elasticsearch/create [_ d] (map->CreateIndexAction d))

;; ----------------------------------------------------------------------
;; Document API

;; Get:

(defrecord GetDocumentAction [name index type id-path]
  i/IntoInterceptor
  (-interceptor [_]
    (assert (vector? id-path) "id-path must be a vector into the context")
    (actions/dynamic-interceptor
     name
     {:enter
      (fn [{:keys [::client] :as context}]
        (let [type (or type default-type)
              id   (str (get-in context id-path))
              req  {:url    [index type id]
                    :method :get}]
          (log/info :msg "GET document" :index index :type type :id id)
          (assoc context :response (es/request client req))))

      :error
      (fn [context exception]
        (log/error :msg "Failed to GET document"
                   :id (get-in context id-path)
                   :status (error-status exception)
                   :error (error-body exception))
        (assoc context :response (error-response exception)))})))

;; Put:

(defrecord PutDocumentAction [name index type id-path body-path]
  i/IntoInterceptor
  (-interceptor [_]
    (actions/dynamic-interceptor
     name
     {:enter
      (fn [{:keys [::client] :as context}]
        (let [type (or type default-type)
              id   (str (get-in context id-path))
              body (get-in context body-path)
              req  {:url    [index type id]
                    :body   body
                    :method :put}]
          (log/info :msg "PUT document" :index index :type type :id id)
          (assoc context :response (es/request client req))))

      :error
      (fn [context exception]
        (log/error :msg "Failed to PUT document"
                   :id (get-in context id-path)
                   :status (error-status exception)
                   :error (error-body exception))
        (assoc context :response (error-response exception)))})))

;; Delete:


(defrecord DeleteDocumentAction [name index type id-path]
  i/IntoInterceptor
  (-interceptor [_]
    (actions/dynamic-interceptor
     name
     {:enter
      (fn [{:keys [::client] :as context}]
        (let [type     (or type default-type)
              id       (str (get-in context id-path))
              request  {:url    [index type id]
                        :method :delete}]
          (log/info :msg "DELETE document" :index index :type type :id id)
          (assoc context :response (es/request client request))))

      :error
      (fn [context exception]
        (log/error :msg "Failed to DELETE document"
                   :id (get-in context id-path)
                   :status (error-status exception)
                   :error (error-body exception)
                   :exception exception)
        (assoc context :response (error-response exception)))})))


;; Literals:

(defmethod f/literal 'vase.elasticsearch/get-document [_ d] (map->GetDocumentAction (vf/with-name d)))
(defmethod f/literal 'vase.elasticsearch/put-document [_ d] (map->PutDocumentAction (vf/with-name d)))
(defmethod f/literal 'vase.elasticsearch/delete-document [_ d] (map->DeleteDocumentAction (vf/with-name d)))

;; ----------------------------------------------------------------------
;; Search API

;; Search:

(defrecord SearchAction [name index params method body]
  i/IntoInterceptor
  (-interceptor [_]
    (actions/dynamic-interceptor
     name
     {:enter
      (request-expr {:url    [index :_search]
                     :method (or method :get)
                     :params params
                     :body   body})

      :error
      (fn [context exception]
        (log/error :msg "Exception during search"
                   :status (error-status exception)
                   :error (error-body exception)
                   :exception exception)
        (assoc context :response (error-response exception)))})))

;; Count:

(defrecord CountAction [name index params method body]
  i/IntoInterceptor
  (-interceptor [_]
    (actions/dynamic-interceptor
     name
     {:enter
      (request-expr {:url    [index :_count]
                     :method (or method :get)
                     :params params
                     :body   body})

      :error
      (fn [context exception]
        (log/error :msg "Exception during count"
                   :status (error-status exception)
                   :error (error-body exception)
                   :exception exception)
        (assoc context :response (error-response exception)))})))

;; Literals:

(defmethod f/literal 'vase.elasticsearch/search [_ d] (map->SearchAction (vf/with-name d)))
(defmethod f/literal 'vase.elasticsearch/count [_ d] (map->CountAction (vf/with-name d)))









(comment

  (def printerceptor
    {:name ::printerceptor
     :enter
     (fn [ctx]
       (prn ctx)
       ctx)})

  (clojure.pprint/pprint
   (request-expr '{:url    [:_search]
                   :method :get
                   :params [q]
                   :body   {:query {:multi_match {:query  (or q "")
                                                  :fields ["title" "overview"]}}}}))


  (require 'qbits.spandex.spec)
  (require '[clojure.spec.test.alpha :as stest])
  (stest/instrument (stest/instrumentable-syms 'qbits.spandex.spec))
  (def c (es/client
          {:hosts       ["http://localhost:9201"]
           :http-client {:basic-auth {:user "elastic" :password ""}}}))


  (require '[io.pedestal.interceptor.chain :as chain])
  (chain/execute {} (map i/interceptor
                         [(f/literal 'vase.elasticsearch/connect {:hosts ["http://localhost:9200"]})
                          (f/literal 'vase.elasticsearch/delete-document {:index "test" :type "_doc" :id-path [:a]})
                          #_(f/literal 'vase.elasticsearch/create {:name "tmdb2" :settings "dev/settings.json"})]))

  ;; Testing a running vase system
  (require '[com.cognitect.vase.main])
  (require '[io.pedestal.http :as http])
  (def server (com.cognitect.vase.main/-main  "dev/tmdb.fern"))

  )
